main.tf

```diff
@@ -158,21 +201,58 @@ resource "aws_cloudwatch_metric_alarm" "cpu_utilization" {
   tags               = var.tags
 }
 
+resource "aws_cloudwatch_metric_alarm" "cpu_utilization_anomaly_detection" {
+  depends_on = [aws_sns_topic_subscription.marbot]
+  count      = (local.cpu_utilization_threshold < -1.5 && local.enabled) ? 1 : 0
+
+  alarm_name          = "marbot-ec2-instance-cpu-utilization-${random_id.id8.hex}"
+  alarm_description   = "${local.alarm_description_prefix}Average CPU utilization over last 10 minutes unexpected. (created by marbot)"
+  evaluation_periods  = local.cpu_utilization_evaluation_periods
+  comparison_operator = "GreaterThanUpperThreshold"
+  threshold_metric_id = "e1"
+  alarm_actions       = [local.topic_arn]
+  ok_actions          = [local.topic_arn]
+  treat_missing_data  = "notBreaching"
+  tags                = var.tags
+
+  metric_query {
+    id          = "e1"
+    expression  = "ANOMALY_DETECTION_BAND(m1)"
+    label       = "CPUUtilization (expected)"
+    return_data = "true"
+  }
+
+  metric_query {
+    id          = "m1"
+    return_data = "true"
+    metric {
+      metric_name = "CPUUtilization"
+      namespace   = "AWS/EC2"
+      period      = local.cpu_utilization_period
+      stat        = "Average"
+
+      dimensions = {
+        InstanceId = var.instance_id
+      }
+    }
+  }
+}
+
 
 
 resource "aws_cloudwatch_metric_alarm" "cpu_credit_balance" {
```

Comment:
An CloudWatch alarm with anomaly detection adds additional costs: $0.30 per alarm/month instead of $0.10.

I'd say CPU Credit Balance, EBS IO Credit Balance do not benefit from anomaly detection alarms. As performance drops significantly after running out of credits, I'd say an alarm based on a fixed threshold should be fine.

What do you think @michaelwittig?

Reply:
I agree and I thought about this as well. I ended up supporting it since we don't turn it on by default? No one is forced to use anomaly detection on those metrics?